### ISSUE TITLE
unitwise_norm fails for 3D convolutions

### ISSUE THREAD

**Issue by froody** [google-deepmind/optax#906]
`unitwise_norm`, used by `adaptive_grad_clip`, only supports a few values of `ndim`, and raises ValueError when applied to a conv3d kernel since `ndim=5` (HWDIO). Would it be acceptable to add an optional `axis` kwarg to `adaptive_grad_clip` and `unitwise_norm`? This would allow specifying the reduction axes at the callsite instead of baking every possible combination into the implementation of `unitwise_norm`.

I'm happy to submit a PR

**Comment by vroulet** [comment]
Hello @froody,

Good catch. The behavior of `adaptive_grad_clip` hides indeed some logic that could mislead users indeed. 
If you are willing to do a pr to let this function handle `ndim=5` that would be great. I don't know exactly how you can add an `axis` and keep the current default behavior, I let you try and see :)

Thank you !

**Comment by aymuos15** [comment]
Hi @vroulet 

I tried having a go at this below. Do you think this is sufficient to kick off a PR or am I missing something? Thanks :)

```diff
diff --git a/optax/transforms/_clipping.py b/optax/transforms/_clipping.py
index eebf6aa..cbb9156 100644
--- a/optax/transforms/_clipping.py
+++ b/optax/transforms/_clipping.py
@@ -18,6 +18,8 @@ Note that complex numbers are also supported, see
 https://gist.github.com/wdphy16/118aef6fb5f82c49790d7678cf87da29
 """
 
+from typing import Optional, Union
+
 import chex
 import jax
 import jax.numpy as jnp
@@ -241,9 +243,21 @@ def per_example_layer_norm_clip(
   )
 
 
-def unitwise_norm(x: chex.Array) -> chex.Array:
-  """Computes norms of each output unit separately."""
-  if jnp.squeeze(x).ndim <= 1:  # Scalars and vectors
+def unitwise_norm(x: chex.Array, axis: Optional[Union[int, tuple[int, ...]]] = None) -> chex.Array:
+  """Computes norms of each output unit separately.
+
+  Args:
+    x: Input array to compute norms for.
+    axis: Axis or axes along which to compute the norm. If None, uses default
+      behavior based on input dimensions.
+
+  Returns:
+    Array of the same shape as x with unit-wise norms.
+  """
+  if axis is not None:
+    # Use provided axes for reduction
+    squared_norm = jnp.sum(numerics.abs_sq(x), axis=axis, keepdims=True)
+  elif jnp.squeeze(x).ndim <= 1:  # Scalars and vectors
     squared_norm = jnp.sum(numerics.abs_sq(x), keepdims=True)
   # Note that this assumes parameters with a shape of length 3 are multihead
   # linear parameters--if you wish to apply AGC to 1D convs, you may need
@@ -254,7 +268,8 @@ def unitwise_norm(x: chex.Array) -> chex.Array:
     squared_norm = jnp.sum(numerics.abs_sq(x), axis=(0, 1, 2), keepdims=True)
   else:
     raise ValueError(
-        f'Expected parameter with shape in {1, 2, 3, 4}, got {x.shape}.'
+        f'Expected parameter with shape in {1, 2, 3, 4}, got {x.shape}. '
+        'Use axis parameter to specify reduction axes for other shapes.'
     )
   chex.assert_is_broadcastable(squared_norm.shape, x.shape)
   return jnp.broadcast_to(jnp.sqrt(squared_norm), x.shape)
@@ -275,13 +290,18 @@ def unitwise_clip(
 
 
 def adaptive_grad_clip(
-    clipping: float, eps: float = 1e-3
+    clipping: float,
+    eps: float = 1e-3,
+    axis: Optional[Union[int, tuple[int, ...]]] = None,
 ) -> base.GradientTransformation:
   """Clips updates to be at most ``clipping * parameter_norm``, unit-wise.
 
   Args:
     clipping: The maximum allowed ratio of update norm to parameter norm.
     eps: An epsilon term to prevent clipping of zero-initialized params.
+    axis: Axis or axes along which to compute the unit-wise norm. If None, uses
+      default behavior based on input dimensions. This is useful for custom
+      parameter shapes like Conv3D (ndim=5).
 
   Returns:
     A :class:`optax.GradientTransformation` object.
@@ -294,7 +314,9 @@ def adaptive_grad_clip(
   def update_fn(updates, state, params):
     if params is None:
       raise ValueError(base.NO_PARAMS_MSG)
-    g_norm, p_norm = jax.tree.map(unitwise_norm, (updates, params))
+    g_norm, p_norm = jax.tree.map(
+        lambda x: unitwise_norm(x, axis=axis), (updates, params)
+    )
     # Maximum allowable norm.
     max_norm = jax.tree.map(lambda x: clipping * jnp.maximum(x, eps), p_norm)
     # If grad norm > clipping * param_norm, rescale.
```

Tests:

```diff
diff --git a/optax/transforms/_clipping_test.py b/optax/transforms/_clipping_test.py
index 573df15..8b5c5e5 100644
--- a/optax/transforms/_clipping_test.py
+++ b/optax/transforms/_clipping_test.py
@@ -39,7 +39,8 @@ class ClippingTest(absltest.TestCase):
     # For a sufficiently high delta the update should not be changed.
     clipper = _clipping.clip(1e6)
     clipped_updates, _ = clipper.update(updates, None)
-    chex.assert_trees_all_close(clipped_updates, clipped_updates)
+    chex.assert_trees_all_close(clipped_updates, updates)
+
     # Clipping at delta=1 should make all updates exactly 1.
     clipper = _clipping.clip(1.0)
     clipped_updates, _ = clipper.update(updates, None)
@@ -73,26 +74,121 @@ class ClippingTest(absltest.TestCase):
       updates_step, _ = clipper.update(self.per_step_updates, None)
       chex.assert_trees_all_close(updates, updates_step)
 
+  def test_adaptive_grad_clip_with_axis(self):
+    """Test adaptive_grad_clip with custom axis parameter for high-dimensional tensors."""
+    
+    # Test case 1: 5D Conv3D kernel (H, W, D, in_channels, out_channels)
+    conv3d_grad = jnp.ones((2, 2, 2, 3, 4)) * 2.0  # Shape: (2,2,2,3,4)
+    conv3d_param = jnp.ones((2, 2, 2, 3, 4))
+    
+    # Test case 2: Regular 2D weight matrix for comparison
+    linear_grad = jnp.ones((5, 6)) * 3.0
+    linear_param = jnp.ones((5, 6))
+    
+    # For Conv3D, we want to compute norms over spatial dimensions (0,1,2)
+    # This gives us per-channel-pair norms of shape (1,1,1,3,4)
+    agc_conv3d = _clipping.adaptive_grad_clip(clipping=0.5, axis=(0, 1, 2))
+    
+    # For linear layer, use default behavior (should work with 2D)
+    agc_linear = _clipping.adaptive_grad_clip(clipping=0.5)
+    
+    # Test Conv3D clipping
+    clipped_conv3d, _ = agc_conv3d.update([conv3d_grad], None, [conv3d_param])
+    clipped_conv3d = clipped_conv3d[0]
+    
+    # Test linear layer clipping  
+    clipped_linear, _ = agc_linear.update([linear_grad], None, [linear_param])
+    clipped_linear = clipped_linear[0]
+    
+    # Verify shapes are preserved
+    self.assertEqual(clipped_conv3d.shape, conv3d_grad.shape)
+    self.assertEqual(clipped_linear.shape, linear_grad.shape)
+    
+    # Verify AGC constraint: ||clipped_grad|| <= clipping * ||param|| (unit-wise)
+    conv3d_grad_norm = _clipping.unitwise_norm(clipped_conv3d, axis=(0, 1, 2))
+    conv3d_param_norm = _clipping.unitwise_norm(conv3d_param, axis=(0, 1, 2))
+    
+    linear_grad_norm = _clipping.unitwise_norm(clipped_linear)
+    linear_param_norm = _clipping.unitwise_norm(linear_param)
+    
+    # Check AGC constraint (with small tolerance for numerical precision)
+    max_allowed_conv3d = 0.5 * conv3d_param_norm + 1e-6
+    max_allowed_linear = 0.5 * linear_param_norm + 1e-6
+    
+    self.assertTrue(jnp.all(conv3d_grad_norm <= max_allowed_conv3d))
+    self.assertTrue(jnp.all(linear_grad_norm <= max_allowed_linear))
+    
+    # Test that without clipping needed, gradients are unchanged
+    small_conv3d_grad = jnp.ones((2, 2, 2, 3, 4)) * 0.1  # Small gradient
+    small_conv3d_param = jnp.ones((2, 2, 2, 3, 4))
+    
+    agc_no_clip = _clipping.adaptive_grad_clip(clipping=2.0, axis=(0, 1, 2))
+    unclipped, _ = agc_no_clip.update([small_conv3d_grad], None, [small_conv3d_param])
+    
+    # Should be nearly unchanged since clipping=2.0 is large
+    chex.assert_trees_all_close(unclipped[0], small_conv3d_grad, rtol=1e-6)
+
+  def test_unitwise_norm_with_axis(self):
+    """Test unitwise_norm function with custom axis parameter."""
+    
+    # Test 5D tensor
+    x = jnp.arange(2*2*2*3*4).reshape(2, 2, 2, 3, 4).astype(jnp.float32)
+    
+    # Test different axis configurations
+    norm_spatial = _clipping.unitwise_norm(x, axis=(0, 1, 2))  # Over spatial dims
+    norm_channels = _clipping.unitwise_norm(x, axis=(3, 4))     # Over channel dims
+    norm_last = _clipping.unitwise_norm(x, axis=-1)             # Over last dim only
+    
+    # Verify shapes
+    self.assertEqual(norm_spatial.shape, x.shape)  # Should broadcast back
+    self.assertEqual(norm_channels.shape, x.shape)
+    self.assertEqual(norm_last.shape, x.shape)
+    
+    # Test that default behavior still works for supported shapes
+    x_2d = jnp.array([[1., 2.], [3., 4.]])
+    norm_default = _clipping.unitwise_norm(x_2d)
+    norm_explicit = _clipping.unitwise_norm(x_2d, axis=None)
+    
+    chex.assert_trees_all_close(norm_default, norm_explicit)
+    
+    # Test error case: unsupported shape without axis
+    x_6d = jnp.ones((2, 2, 2, 2, 2, 2))
+    with self.assertRaises(ValueError):
+      _clipping.unitwise_norm(x_6d)  # Should fail without axis
+    
+    # But should work with axis specified  
+    norm_6d = _clipping.unitwise_norm(x_6d, axis=(0, 1, 2))
+    self.assertEqual(norm_6d.shape, x_6d.shape)
+
   def test_adaptive_grad_clip(self):
     updates = self.per_step_updates
     params = self.init_params
     for i in range(1, STEPS + 1):
       clip_r = 1.0 / i
-      clipper = _clipping.adaptive_grad_clip(clip_r)
+      clipper = _clipping.adaptive_grad_clip(clipping=clip_r)
 
-      # Check that the clipper actually works and upd_norm is < c * param_norm.
+      # Check that the clipper actually works and upd_norm is < clip_r * param_norm.
       updates, _ = clipper.update(updates, None, params)
       u_norm, p_norm = jax.tree.map(_clipping.unitwise_norm, (updates, params))
       cmp = jax.tree.map(
-          lambda u, p, c=clip_r: u - c * p < 1e-6, u_norm, p_norm
+          lambda u, p, c=clip_r: (u - c * p < 1e-6).all(), u_norm, p_norm
       )
       for leaf in jax.tree.leaves(cmp):
-        self.assertTrue(leaf.all())
+        self.assertTrue(leaf)
 
       # Check that continuously clipping won't cause numerical issues.
       updates_step, _ = clipper.update(self.per_step_updates, None, params)
       chex.assert_trees_all_close(updates, updates_step)
 
+    # Simple 2D tensor case for default behavior
+    grads_small = [jnp.array([[1., 2.], [3., 4.]])]
+    params_small = [jnp.ones_like(grads_small[0])]
+    agc_default = _clipping.adaptive_grad_clip(clipping=1.0)
+    clipped_small, _ = agc_default.update(grads_small, None, params_small)
+    clipped_norm_small = _clipping.unitwise_norm(clipped_small[0])
+    param_norm_small = _clipping.unitwise_norm(params_small[0])
+    self.assertTrue(jnp.all(clipped_norm_small <= 1.0 * param_norm_small + 1e-6))
+
   def test_per_example_global_norm_clip(self):
     grads = [  # 3 users, 2 components
         jnp.array([
```

**Comment by vroulet** [comment]
It looks good, you can start a PR! @rajasekharporeddy @selamw1 would you be able to then take a look at that PR? Thanks!

**Comment by aymuos15** [comment]
Thank you very much! Have created one.

**Comment by rdyro** [comment]
Closing as fixed
